### ISSUE TITLE
Unclear driver updare protocol

### ISSUE THREAD

**Issue by krOoze** [KhronosGroup/Vulkan-Docs#1319]
These days driver updates seems to be fine to happen on runtime without restart. But I see many apps misbehaving. Maybe for the sake of general quality of SW, this case should be handled properly.

Perhaps the spec should be more clear about what is the protocol when a driver is updated. Maybe even add some things to facilitate it.

There seems to be a `VK_ERROR_DEVICE_LOST` for a `VkDevice` and `VkPhysicalDevice`. For the second one the spec seems to imply it is probably unrecoverable. On the other hand, I would expect driver update to cause exactly that `VkPhysicalDevice` to become lost. Or is the same `VkPhysicalDevice` guaranteed after driver update?

It is probably fine if an app panics on thermal problem, GSOD, or the GPU gets hit by a meteor. On the other hand there is no way to disambiguate the normal and good case of a driver update. And the driver update may take a long time, which does look somewhat like some permanent problem of the PC. So I think there is some guidance missing here how to make a well-behaved resilient SW.

Also, driver update might update the `vulkan-1.dll`. I guess that means statically loaded app is stuck with the old version, which should be OK, except any bugs in the Loader and missing support for new extensions and new Vulkan versions. For manually loading app, there should perhaps be guidance when to recreate the `VkInstance` as well (which does not really have `VK_ERROR_INSTANCE_LOST`) and when to reload the dll.

Somewhat relatedly, I think there is missing support for hypothetical hotpluggable GPUs. I think `vkEnumeratePhysicalDevices` is immutable (at least that is the default, and descriptions of the command do not say otherwise). There do not seem to be any event to report a new device connected. And it seems one should not use `vkCreateInstance` frivolously, as it seems to be relatively expensive on some drivers (so probably not wise to poll for the information).

**Comment by lenny-lunarg** [comment]
You're correct that this is a problem. I've given my thoughts below:

> Also, driver update might update the vulkan-1.dll. I guess that means statically loaded app is stuck with the old version

We have a change that we've been working on which should alleviate this, but it's not merged at the moment. Honestly though, needing to restart an application to get a driver update seems like a reasonable requirement unless you support hotplugging, which we really don't. This brings me to the other point, which I think is the more involved one:

> Somewhat relatedly, I think there is missing support for hypothetical hotpluggable GPUs. I think vkEnumeratePhysicalDevices is immutable (at least that is the default, and descriptions of the command do not say other

This is actually a rather intricate issue, but it's definitely something where Vulkan is lacking right now. There is highly limited support for hotplugging but it really isn't a viable solution to hotplugging. The loader is capable of picking up new GPUs when you call `vkEnumeratePhysicalDevices`, but it can't pick up new drivers. This means if your hotplugged GPU is from the same vendor, you might (but are not guaranteed to) see the new device. If the hotplugged GPU is from a different vendor, you won't see it. You can get a new GPU by creating a new instance. This is a more invasive change which probably isn't acceptable for most applications that want to support hotplugging. On top of that, we have no way to actually check if a new GPU has been added, and creating a new instance just to call `vkEnumeratePhysicalDevices` really isn't a viable solution.

As far as fixing that issue, it's something we've been aware of for a while, but haven't been motivated enough to fix it yet. For games, it's generally not an issue because no game that I've seen has any interest in trying to move from one GPU to another. This case is far more relevant for applications like CAD software, where it would be much more normal for someone to have a laptop that they can plug into a workstation for renders. But Vulkan isn't really used in CAD software as of today. So the result of all this is that we know this is an issue, but it's a big job and we've had higher priority issues. I imagine that we'll be motivated to work on this eventually, but I'm definitely not willing to promise any sort of timeframe. I suspect it would be motivated by when Vulkan users actually consider it to be a problem in practice.

**Comment by jjulianoatnv** [comment]
There are issues at multiple layers.
1. The "vulkan loader" needs some code changes to support the scenarios under discussion.
1. I believe new Vulkan API is needed for communicating state change to the app. This opinion might not be shared by everyone.
1. I predict that Vulkan apps will have to actively participate through the new Vulkan API.
1. There might be need to change the interface that exists between the Vulkan loader and Vulkan ICDs. I'm unclear if this is true, or if the existing interface is good enough. There was relatively-recent change to this interface, and it might have gone far enough, or additional change might still be needed.

In addition to the specific issues that @krOoze enumerated, there are additional specific issues that I believe can also be addressed by the same future solution. Within Khronos I have in the past described a potential approach to fixing the family of issues. We have multiple internal issues tracking pieces of this. It has been a while since they have received attention within Khronos. Thank you for raising the topic here on github.

**Comment by krOoze** [comment]
> Honestly though, needing to restart an application to get a driver update seems like a reasonable requirement

I guess `vulkan-1.dll` itself is too much to hope for. It is little bit a problem with `lib`+`dll` generally.
But seems doable for uber-explicit app, that just loads `vkGetInstanceProcAddr` directly from `vulkan-1.dll`. It seems it is just needed to tell such app to do it in the spec when to close the `dll` and load it again.

In other cases (only the ICD update) honestly it is kinda annoying. Imagine you would have to restart a word processor when you have installed a new printer. When LPT printer was a thing, people would say "of course you have to restart your whole PC". Nowadays we have different standards and everyone would cosider that dumb to have to do that.

Maybe I should mention some usecases here that I imagine could do better:

- Browser permanently open with some work in 50 tabs. Maybe watching a YouTube video. (Technically this is not necesserily a Vulkan SW, but to illustrate)
  - Current state: everything that is Flash or Media or perhaps loading a page while using accellerated rendering glitches and needs browser restart after driver update (without reporting anything to the user). User posibly loses any progress (e.g. the position of the YouTube video) after restart.
  - Ideal state: after the driver update finishes, everything that needs GPU API reinitializes itself and then continues like nothing ever happened.
 - Game is minimized. User paused it with intention of continuing at some point later.
   - Current state: the game crashes, freezes, or glitches; possibly losing progress.
   - Ideal state: after driver update finishes, loading screen is triggered to reinitialize things. Then on, stuff is exactly as user left it. If the game was minimized the whole time, it would be completely seamless.

> The loader is capable of picking up new GPUs when you call `vkEnumeratePhysicalDevices`,

Well then that currently seem to be non-conformant behavior per the default immutability of results by Vulkan.

> I suspect it would be motivated by when Vulkan users actually consider it to be a problem in practice.

Well, that is bit of a chicken-egg problem. (Possibly inexperienced) end users do not know they should want this. And game makers are interested only in what end users are interested in.

External hotpluggable GPUs are equally a chicken-egg problem. Who would want to invest in some solution and make it a thing, if they know it will take another two years to get some proper support and meanwhile all the old apps and games would not work right on it if it needs changes to the API.

Also there is the other weirdness on notebooks, when the driver is there, but the GPU is actually powered down. That is already something like a hotplugability that is even widely used even now. Feels like it needs some consistent design\API instead of just piling hacks to make things work "when they become a problem".

**Comment by jjulianoatnv** [comment]
@krOoze , you are correct that there are several potential use cases for this after it has been fixed. You did not name: thunderbolt-attached GPU that is connected post-boot, and a few other cases.

I personally have strong desire for a fix for this. Professional apps, web browsers, desktop/HMD compositors, and other non-toy scenarios will need this to be addressed. Priorities withing Khronos have not aligned yet to give this issue more attention than it has received.

**Comment by krOoze** [comment]
I enumerated it by anything but name. Anyway my main concern for now is general quality of apps, as it impacts everyone.

My issue is also with "when it becomes a problem". It is good approach for devices (and their features), but bad approach for infrastructure and ecosystem. It leads to ugly things like `vkEnumerateInstanceVersion`(, which could have been in 1.0, and meanwhile does not even look ready for major version updates.).

**Comment by jjulianoatnv** [comment]
> My issue is also with "when it becomes a problem".

FWIW, I agree with you on this point.
